src/core/tsi/ssl_transport_security_util.h

```diff
@@ -0,0 +1,144 @@
+/*
+ *
+ * Copyright 2015 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#ifndef GRPC_CORE_TSI_SSL_TRANSPORT_SECURITY_UTIL_H_
+#define GRPC_CORE_TSI_SSL_TRANSPORT_SECURITY_UTIL_H_
+
+#include <grpc/support/port_platform.h>
+
+#include <openssl/x509.h>
+
+#include "absl/strings/string_view.h"
+
+#include <grpc/grpc_security_constants.h>
+
+#include "src/core/tsi/ssl/key_logging/ssl_key_logging.h"
+#include "src/core/tsi/transport_security_interface.h"
+
+// Converts an SSL error status code to a readable string.
+//
+// error: the SSL error status code.
+//
+// return: the corresponding status string.
+const char* ssl_error_string(int error);
+
+// Logs the SSL error stack.
+void log_ssl_error_stack(void);
+
+// Performs an SSL_write and handle errors.
+//
+// ssl: the SSL object to write to.
+// unprotected_bytes: the buffer containing the bytes for writing to |ssl|.
+// unprotected_bytes_size: the size of the buffer |unprotected_bytes|.
+//
+// return: TSI_OK if the write operation succeeds or corresponding TSI errors.
+tsi_result do_ssl_write(SSL* ssl, unsigned char* unprotected_bytes,
+                        size_t unprotected_bytes_size);
+
+// Performs an SSL_read and handle errors.
+//
+// ssl: the SSL object to read from.
+// unprotected_bytes: the buffer to which this function will populate the read
+//                    result from |ssl|.
+// unprotected_bytes_size: the maximum size of the buffer |unprotected_bytes|.
+//                         This will be populated with the size of the bytes
+//                         read from |ssl| if this function returns TSI_OK.
+//
+// return: TSI_OK if the write operation succeeds or corresponding TSI errors.
+tsi_result do_ssl_read(SSL* ssl, unsigned char* unprotected_bytes,
+                       size_t* unprotected_bytes_size);
+
+// Builds a maximum-size TLS frame if there is enough (|buffer_offset| +
+// |unprotected_bytes_size| >= |buffers_size|) data. Otherwise it copies the
```

Comment:
I am not super familiar with this function, just asking from the API's perspective:
if offset + unprotected_size > total_size, doesn't that mean we've reached the upper limit that this buffer can hold? Why in that case can we still build the TLS frame? Did I get something wrong here?

Reply:
The total_size in your comment refers to the size that's needed to build a TLS frame. Let me explain it by an example: say we need 10 bytes of data to build a TLS frame (so the buffer_size is 10) and there's already 8 bytes in the buffer (offset = 8). The user/caller passes another 7 bytes of data (input data size = 7) into `ssl_protector_protect_util` function. Then this function will take 2 bytes and send 10 bytes to build a TLS frame. It also updates the `unprotected_bytes_size` with the bytes consumed (i.e., 2). So the caller knows that 2 bytes out of the 7 bytes that the caller passes to the function are consumed. Next time the caller should pass the remaining 5 bytes, and this function will copy all of them to the inner buffer.